src/node_main_instance.cc

```diff
@@ -208,27 +200,18 @@ std::unique_ptr<Environment> NodeMainInstance::CreateMainEnvironment(
   CHECK(!context.IsEmpty());
   Context::Scope context_scope(context);
 
-  std::unique_ptr<Environment> env = std::make_unique<Environment>(
+  DeleteFnPtr<Environment, FreeEnvironment> env { CreateEnvironment(
       isolate_data_.get(),
       context,
       args_,
       exec_args_,
-      static_cast<Environment::Flags>(Environment::kIsMainThread |
-                                      Environment::kOwnsProcessState |
-                                      Environment::kOwnsInspector));
-  env->InitializeLibuv(per_process::v8_is_profiling);
-  env->InitializeDiagnostics();
-
-  // TODO(joyeecheung): when we snapshot the bootstrapped context,
-  // the inspector and diagnostics setup should after after deserialization.
```

Comment:
Why is this removed? I am also not convinced that it's a good idea to use `CreateEnvironment()` and `LoadEnvironment()` on the main thread at this point, at least that's not something we can guarantee at the moment - to snapshot the bootstrap process incrementally, we need to insert logic during what's being done in those two calls, so we'll have to revert this soon either way so that a) `RunBootstrapping()` can be broken into two phases, the first of which of produce a snapshot and b) `NodeMainInstance::CreateMainEnvironment` is able to either continue bootstrapping from a snapshot, or bootstrap from scratch. This cannot be done if we require the bootstrapping to be done in one go in `CreateEnvironment()`

Reply:
> Why is this removed?

Because this PR does make it so that inspector + diagnostics are setup after deserialization, in `LoadEnvironment()` – or am I not understanding the `TODO` correctly?

> I am also not convinced that it's a good idea to use `CreateEnvironment()` and `LoadEnvironment()` on the main thread at this point, at least that's not something we can guarantee at the moment

I don’t think we have to guarantee that, either, but is there a reason not to do it for now?

> to snapshot the bootstrap process incrementally, we need to insert logic during what's being done in those two calls, so we'll have to revert this soon either way so that a) `RunBootstrapping()` can be broken into two phases, the first of which of produce a state ready to be snapshotted and b) `NodeMainInstance::CreateMainEnvironment` is able to either continue bootstrapping from a snapshot, or bootstrap from scratch. This cannot be done if we require the bootstrapping to be done in one go in `CreateEnvironment()`, and is not something that should be exposed because it's just a temporary thing and the breaking point of `RunBootstrapping()` would be arbitrary depending on the progress of the snapshot effort.

Is there a reason why we can’t move the parts of bootstrapping that we want to snapshot into `CreateEnvironment()`? My understanding was that that was kind of the goal.

Reply:
> I don’t think we have to guarantee that, either, but is there a reason not to do it for now?

I think it's OK to do it for now, but I can't see an upside of it if it's not going to be guaranteed, and the downside is if you want to look into the commit history to understand how the sequence changes, you'll need to jump to the definition of  `CreateEnvironment ()` at some point and then going back to see them inline again in the next commit.

> Is there a reason why we can’t move the parts of bootstrapping that we want to snapshot into CreateEnvironment()? My understanding was that that was kind of the goal.

First and foremost, `CreateEnvironment()` currently takes an existing `Context`, while when we have snapshot support, the context does not exist until it is deserialized from the snapshot. Also since we need to have a set of C++ address first to hook them up with the JS to C++ references in the snapshot, we will need to allocate a chunk of memory for `Environment` first, deserialize the Context with the help of some additional data and code (https://github.com/joyeecheung/node/blob/ae98b7178b3c1306b93b2f1613450cda4d7eb580/src/node_main_instance.cc#L221-L230), and then use replacement new to initialize that memory, I don't think we can hide this all from the signature of CreateEnvironment while the project is WIP.

Reply:
@joyeecheung I guess for me the question is more like, “can we keep supporting the current definition of `CreateEnvironment()` for at least one major Node.js version”, and I feel like the answer to that is yes. I don’t think this is going to stand in the way of future changes in any way, including possibly moving the main thread away from this API.

Reply:
> Because this PR does make it so that inspector + diagnostics are setup after deserialization, in LoadEnvironment() – or am I not understanding the TODO correctly?

This is more like an invariant that we need to keep - so it's probably not entirely a TODO, but I think it should be at least kept in the code as there is only "done" of this when there is no need to pay attention to this invariant (but it should be kept nonetheless IMO)

Reply:
@joyeecheung Hm … in that case I’d suggest adding explicit `CHECK`s that make sure that libuv/diagnostics initialization have not already run to the start of `LoadEnvironment()`?